crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -408,17 +344,23 @@ impl<'db> VisibilityConstraintsBuilder<'db> {
             .or_insert_with(|| self.interiors.push(node))
     }
 
-    /// Adds a new visibility constraint that checks a single [`Constraint`]. Provide different
-    /// values for `copy` if you need to model that the constraint can evaluate to different
-    /// results at different points in the execution of the program being modeled.
+    /// Adds a new visibility constraint that checks a single [`Constraint`].
+    ///
+    /// [`ScopedConstraintId`]s are the “variables” that are evaluated by a TDD. A TDD variable has
+    /// the same value no matter how many times it appears in the ternary formula that the TDD
+    /// represents.
+    ///
+    /// However, we sometimes have to model how a `Constraint` can have a different runtime
+    /// value at different points in the execution of the program. To handle this, you can take
+    /// advantage of the fact that the [`Constraints`] arena does not deduplicate `Constraint`s.
+    /// You can add a `Constraint` multiple times, yielding different `ScopeConstraintId`s, which
+    /// you can then create separate TDD atoms for.
     pub(crate) fn add_atom(
         &mut self,
-        constraint: Constraint<'db>,
-        copy: u8,
+        constraint: ScopedConstraintId,
```

Comment:
Using `ScopedConstraintId` here means we have to be sure that a `VisibilityConstraints` is also scoped to a single... scope. This is the case, and there seems little chance it would ever not be the case... but it might be worth a doc comment mention anyway?